### PR TITLE
doc(iam): add samples for mocks and StatusOr usage

### PR DESCRIPTION
**Pull request by scotthart**
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/6957)

### CODE REVIEW

**Comment by coryan** [comment]
Why is this `chore(iam)`?  Adding documentation and examples is valuable work, it helps our customers.  This should be `doc(iam)`.

File: google/cloud/iam/samples/iam_samples.cc [hunk 1]
```diff
@@ -27,6 +27,35 @@
...
+void ExampleStatusOr(std::vector<std::string> const& argv) {
+  //! [example-status-or]
+  namespace iam = ::google::cloud::iam;
+  [](std::string const& project_id) {
+    iam::IAMClient client(iam::MakeIAMConnection());
+    int count = 0;
+    // The actual type of `service_account` is
+    // google::cloud::StatusOr<google::iam::admin::v1::ServiceAccount>, but
+    // we expect it'll most often be declared with auto like this.
+    for (auto const& service_account :
```
Comment (coryan) [inline google/cloud/iam/samples/iam_samples.cc:40]: Consider: `auto const& sa`, longer names do not make code more readable.  There are actual studies on that:

https://www.researchgate.net/publication/318797794_Fixing_Faults_in_C_and_Java_Source_Code_Abbreviated_vs_Full-Word_Identifier_Names

File: google/cloud/iam/samples/iam_samples.cc [hunk 2]
```diff
@@ -27,6 +27,35 @@
...
+         client.ListServiceAccounts(absl::StrCat("projects/", project_id))) {
```
Comment (coryan) [inline google/cloud/iam/samples/iam_samples.cc:41]: In examples, try to introduce as few libraries as possible, `"projects/" + project_id` should be familiar to all C++ programmers, `absl::StrCat` may not be.

**Review by coryan — COMMENTED** [review]
I think the examples could be made easier to read.

**Review by scotthart — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/6957)* status: 0 of 6 files reviewed, 2 unresolved discussions (waiting on @coryan)

___
*[google/cloud/iam/samples/iam_samples.cc, line 40 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/6957#-Me6hnh71qzyX0IHBh6l-r1-40:-MeBG-TJ-ghO0uBNdTKP:b88bcrb) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/3efb8366c69fc55df7ed77315621c17b0d13029e/google/cloud/iam/samples/iam_samples.cc#L40)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

Consider: `auto const& sa`, longer names do not make code more readable.  There are actual studies on that:

https://www.researchgate.net/publication/318797794_Fixing_Faults_in_C_and_Java_Source_Code_Abbreviated_vs_Full-Word_Identifier_Names

</blockquote></details>

I feel that this and the style guide are in conflict. I could be wrong, but I would prefer to leave this for now.
___
*[google/cloud/iam/samples/iam_samples.cc, line 41 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/6957#-Me6hnh71qzyX0IHBh6l-r1-41:-MeBFhb15DDDrq-NOW6k:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/3efb8366c69fc55df7ed77315621c17b0d13029e/google/cloud/iam/samples/iam_samples.cc#L41)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

In examples, try to introduce as few libraries as possible, `"projects/" + project_id` should be familiar to all C++ programmers, `absl::StrCat` may not be.
</blockquote></details>

Done.

**Review by coryan — APPROVED** [review]
> Reviewed 5 of 6 files at r1, 1 of 1 files at r2.
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/6957)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @scotthart)

___
*[google/cloud/iam/samples/iam_samples.cc, line 40 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/6957#-Me6hnh71qzyX0IHBh6l-r1-40:-MeBJJZmAWklbJaVFHT5:b-u1xqqk) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/3efb8366c69fc55df7ed77315621c17b0d13029e/google/cloud/iam/samples/iam_samples.cc#L40)):*
> I feel that this and the style guide are in conflict.

First, I don't care what the style guide says.  It is a guide, not a suicide pact. If the example code is *worse* because we follow the style guide, then I will not follow the style guide.

Second, I am not sure why you feel that way. @devjgm already addressed this in: https://github.com/googleapis/google-cloud-cpp/pull/6794#pullrequestreview-689556240

Quoting https://google.github.io/styleguide/cppguide.html#General_Naming_Rules

> Generally speaking, descriptiveness should be proportional to the name's scope of visibility. For example, n may be a fine name within a 5-line function, but within the scope of a class, it's likely too vague.

Anyway, I don't care enough to try to convince you, change it or leave as-is, your call.
___
*[google/cloud/iam/samples/iam_samples.cc, line 41 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/6957#-Me6hnh71qzyX0IHBh6l-r1-41:-MeBJezR-aW1aLWu37Qu:b-to9xn2) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/3efb8366c69fc55df7ed77315621c17b0d13029e/google/cloud/iam/samples/iam_samples.cc#L41)):*
<details><summary><i>Previously, scotthart (Scott Hart) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

You missed this one, please fix.

**Pull request merged**